aggregation/performing-complex-mongodb-data-aggregation-queries-with-go.md

```diff
@@ -0,0 +1,232 @@
+
+```json
+{
+    "_id": ObjectId("5e3b37e51c9d4400004117e6"),
+    "name": "The Polyglot Developer Podcast",
+    "author": "Nic Raboy",
+    "tags": ["development", "programming", "coding"]
+}
+```
+
+So what would our aggregation query look like if we wanted to include the podcast information with the episode information? We might try to do something like this:
+
+```go
+showLoadedCursor, err := episodesCollection.Aggregate(ctx, []bson.D{
+    bson.D{{"$lookup", bson.D{{"from", "podcasts"}, {"localField", "podcast"}, {"foreignField", "_id"}, {"as", "podcast"}}}},
+    bson.D{{"$unwind", bson.D{{"path", "$podcast"}, {"preserveNullAndEmptyArrays", false}}}},
+})
+if err != nil {
+    panic(err)
+}
+var showsLoaded []interface{}
+if err = showLoadedCursor.All(ctx, &showsLoaded); err != nil {
+    panic(err)
+}
+fmt.Println(showsLoaded)
+```
+
+In the above example, we are using the [$lookup](https://docs.mongodb.com/manual/reference/operator/aggregation/lookup/) operator to join from the `podcasts` collection using the `podcast` field in our `episodes` collection and the `_id` field in our foreign `podcasts` collection. The output from the `$lookup` operation will be an array stored as `podcast`.
+
+After the `$lookup` operation, we make use of the [$unwind](https://docs.mongodb.com/manual/reference/operator/aggregation/unwind/) operator to flatten the array that we had previously created. Think of flattening or deconstructing an array as taking an array and now outputting each element of that array as a document in the result set.
+
+If we were to run our aggregation, we might end up with results that look like this:
+
+```plaintext
+[[{_id ObjectID("5e3b381c1c9d4400004117e7")} {podcast [{_id ObjectID("5e3b37e51c9d4400004117e6")} {name The Polyglot Developer Podcast} {author Nic Raboy} {tags 
+[development coding programming]}]} {title Episode #1} {description The first episode} {duration 25}] [{_id ObjectID("5e3b38511c9d4400004117e8")} {podcast [{_id 
+ObjectID("5e3b37e51c9d4400004117e6")} {name The Polyglot Developer Podcast} {author Nic Raboy} {tags [development coding programming]}]} {title Episode #2} {desc
+ription The second episode} {duration 30}]]
+```
+
+Notice that each podcast episode in the above results is now printed with the show information. This saves you from having to execute multiple `Find` operations within your Go code.
+
+The above query that we saw is great, but I think we can do better.
+
+In the first example that used `$lookup` and `$unwind` we were using an `[]interface{}` to work with the results. Not the end of the world, but if we wanted to access particular fields, use an autocomplete, etc., things might get a little messy. Instead, we can create a native Go data structure to represent the results of our aggregation.
+
+```go
+// PodcastEpisode represents an aggregation result-set for two collections
+type PodcastEpisode struct {
```

Comment:
This is a different `PodcastEpisode` definition than the one in the `main.go` file. The one in the code has an additional `PublishDate` field.